resources/component_runit_service.rb

```diff
@@ -0,0 +1,114 @@
+#
+# Cookbook Name:: enterprise
+# Resource:: component_runit_service
+#
+# Copyright:: 2018, Chef Software, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# A runit service wrapper for a chef component within an omnibus project
+#
+resource_name :component_runit_service
+
+# :component is the name of the component/service of the project for which
+# this resource will manage a runit definition
+property :component, String, name_property: true
+
+# :package is the project name and is used as the key into which part of the
+# node attributes will be referenced for package+component configuration
+# because add-ons have different package names
+# e.g. :package => 'reporting' or :package => 'manage'
+property :package, String, default: 'private_chef'
+
+property :log_directory, String # where the runit service will manage itself
+property :svlogd_size, Integer # log rotation : a size limit for log files
+property :svlogd_num, Integer # log rotation : the number of log files to keep
+property :ha, [TrueClass, FalseClass] # is this a highly-available service?
+property :control, Array # control signal overrides for runit and runsv
+
+# :runit_attributes - optionally set other properties on the runit_service
+# resource that are not explicitly supported by this resource
+property :runit_attributes, Hash, default: {}
+
+action :enable do
+  log_directory = new_resource.log_directory || node[new_resource.package][new_resource.component]['log_directory']
+  svlogd_size = new_resource.svlogd_size || node[new_resource.package][new_resource.component]['log_rotation']['file_maxbytes']
+  svlogd_num = new_resource.svlogd_num || node[new_resource.package][new_resource.component]['log_rotation']['num_to_keep']
+
+  # runit resources don't support reloading the log service as an action :(
+  execute "restart_#{new_resource.component}_log_service" do
+    command "#{node['runit']['sv_bin']} restart #{node['runit']['sv_dir']}/#{new_resource.component}/log"
```

Comment:
It would be nice to create new properties on this resource for sv_bin and sv_dir that lazily default to these attributes. Using attributes in a resource is an antipattern, but we're sort of stuck with it from a backwards compatibility standpoint due to the history of this codebase.

Reply:
Yup. Everything that uses this cookbook also uses the `runit` cookbook and liberally depends on `runit`'s attribute namespace to tinker with things. If we update these node attribute value retrievals to live in new property lazy defaults, I can think of only benefit: the resource more clearly communicates to a human reader that it has attribute namespace expectations outside of its own namespace. The technical dependency remains but is maybe isolated to a single, clear call in the property defaults.